valhalla/loki/polygon_search.h

```diff
@@ -0,0 +1,67 @@
+#ifndef VALHALLA_LOKI_POLYGON_SEARCH_H_
+#define VALHALLA_LOKI_POLYGON_SEARCH_H_
+
+#include <boost/geometry.hpp>
+#include <boost/geometry/geometries/register/point.hpp>
+#include <boost/geometry/geometries/register/ring.hpp>
+
+#include <valhalla/baldr/graphreader.h>
+#include <valhalla/midgard/constants.h>
+#include <valhalla/midgard/pointll.h>
+#include <valhalla/proto/options.pb.h>
+#include <valhalla/sif/dynamiccost.h>
+
+#include <functional>
+
+// TODO: register PointLL in its own header, so it's available for bg across valhalla?
+BOOST_GEOMETRY_REGISTER_POINT_2D(valhalla::midgard::PointLL,
+                                 double,
+                                 boost::geometry::cs::geographic<boost::geometry::degree>,
+                                 first,
+                                 second)
+BOOST_GEOMETRY_REGISTER_RING(std::vector<valhalla::midgard::PointLL>)
+static const auto Haversine = [] {
+  return boost::geometry::strategy::distance::haversine<float>(valhalla::midgard::kRadEarthMeters);
+};
+
+namespace valhalla {
+namespace loki {
+
+using line_bg_t = boost::geometry::model::linestring<midgard::PointLL>;
+using ring_bg_t = std::vector<midgard::PointLL>;
+// map of tile for map of bin ids & their ring ids
+using bins_collector =
+    std::unordered_map<uint32_t, std::unordered_map<unsigned short, std::vector<size_t>>>;
+
+/**
+ * Finds all edge IDs which are intersected by the ring
+ *
+ * @param rings The (optionally closed) rings to intersect edges with
+ * @param reader GraphReader instance
+ *
+ */
+std::unordered_set<valhalla::baldr::GraphId>
+edges_in_rings(const std::vector<ring_bg_t>& rings,
```

Comment:
it would be great if this interface were only this single function. here's what im thinking. im not a fan of having the guts of conversion in loki/worker.cc. aside: i know we do that for locations but its not because we wanted to, its because we didnt used to have protobuf when the project first started and we just havent gotten to converting loki::search to work with protobuf natively.

anyway.. what if we change this funciton signature to be:

```c++
std::unordered_set<valhalla::baldr::GraphId>
edges_in_rings(const protobuf::repeated<Ring>& rings, //that typename isnt right i just dont remember right now
               baldr::GraphReader& reader,
               const std::shared_ptr<sif::DynamicCost>& costing,
               float max_length);
```

then this function could do some checks at the beginning and throw if any ring is too crazy large. it can still throw the same valhalla exception. and we can take the other two functions that are here and move them into an anonymous name space in the cc file to remove them from the public interface. i just prefer to encapsulate related functionality as much as we can rather than spread it around. we want to have as small an public interface to each class/namespace etc as possible so it keeps the spider webbing down :smile:

Reply:
Ok understood! 
My reasoning was that these other 2 functions might be useful at some point so I tried to make them a little generic. In case there's other polygon/ring stuff you'd like to care of in the future. But of course it's easily refactored once that's actually the case.

Reply:
yep exactly my thoughts, when it becomes useful we can expand the public interface

Reply:
Done